### ISSUE TITLE
[REVIEW]: The Basic Model Interface 2.0: A standard interface for coupling numerical models in the geosciences

### ISSUE THREAD

**Issue by whedon** [openjournals/joss-reviews#2317]
**Submitting author:** @mdpiper (<a href="http://orcid.org/0000-0001-6418-277X">MARK PIPER</a>)
**Repository:** <a href="https://github.com/csdms/bmi" target ="_blank">https://github.com/csdms/bmi</a>
**Version:** v2.0
**Editor:** @diehlpk
**Reviewer:** @yangbai90, @teuben
**Archive:** <a href="https://doi.org/10.5281/zenodo.3955010" target="_blank">10.5281/zenodo.3955010</a>

**:warning: JOSS reduced service mode :warning:**

Due to the challenges of the COVID-19 pandemic, JOSS is currently operating in a "reduced service mode". You can read more about what that means in [our blog post](https://blog.joss.theoj.org/2020/05/reopening-joss).

## Status

[![status](https://joss.theoj.org/papers/f0700d419c760b528d79468be4346d4d/status.svg)](https://joss.theoj.org/papers/f0700d419c760b528d79468be4346d4d)

Status badge code:

```
HTML: <a href="https://joss.theoj.org/papers/f0700d419c760b528d79468be4346d4d"><img src="https://joss.theoj.org/papers/f0700d419c760b528d79468be4346d4d/status.svg"></a>
Markdown: [![status](https://joss.theoj.org/papers/f0700d419c760b528d79468be4346d4d/status.svg)](https://joss.theoj.org/papers/f0700d419c760b528d79468be4346d4d)
```
**Reviewers and authors**:

Please avoid lengthy details of difficulties in the review thread. Instead, please create a new issue in the <a href="https://github.com/csdms/bmi" target="_blank">target repository</a> and link to those issues (especially acceptance-blockers) by leaving comments in the review thread below. (For completists: if the target issue tracker is also on GitHub, linking the review thread in the issue or vice versa will create corresponding breadcrumb trails in the link target.)

## Reviewer instructions & questions

@yangbai90 & @teuben, please carry out your review in this issue by updating the checklist below. If you cannot edit the checklist please:
1. Make sure you're logged in to your GitHub account
2. Be sure to accept the invite at this URL: https://github.com/openjournals/joss-reviews/invitations

The reviewer guidelines are available here: https://joss.readthedocs.io/en/latest/reviewer_guidelines.html. Any questions/concerns please let @diehlpk know.

✨ **Please try and complete your review in the next six weeks** ✨ 

### Conflict of interest

- [x] I confirm that I have read the [JOSS conflict of interest (COI) policy](https://github.com/openjournals/joss/blob/master/COI.md) and that: I have no COIs with reviewing this work or that any perceived COIs have been waived by JOSS for the purpose of this review.

### Code of Conduct

- [x] I confirm that I read and will adhere to the [JOSS code of conduct](https://joss.theoj.org/about#code_of_conduct).

### General checks

- [x] **Repository:** Is the source code for this software available at the <a target="_blank" href="https://github.com/csdms/bmi">repository url</a>?
- [x] **License:** Does the repository contain a plain-text LICENSE file with the contents of an [OSI approved](https://opensource.org/licenses/alphabetical) software license?
- [x] **Contribution and authorship:** Has the submitting author (@mdpiper) made major contributions to the software? Does the full list of paper authors seem appropriate and complete?

### Functionality

- [x] **Installation:** Does installation proceed as outlined in the documentation?
- [x] **Functionality:** Have the functional claims of the software been confirmed?
- [x] **Performance:** If there are any performance claims of the software, have they been confirmed? (If there are no claims, please check off this item.)

### Documentation

- [x] **A statement of need:** Do the authors clearly state what problems the software is designed to solve and who the target audience is?
- [x] **Installation instructions:** Is there a clearly-stated list of dependencies? Ideally these should be handled with an automated package management solution.
- [x] **Example usage:** Do the authors include examples of how to use the software (ideally to solve real-world analysis problems).
- [x] **Functionality documentation:** Is the core functionality of the software documented to a satisfactory level (e.g., API method documentation)?
- [x] **Automated tests:** Are there automated tests or manual steps described so that the functionality of the software can be verified?
- [x] **Community guidelines:** Are there clear guidelines for third parties wishing to 1) Contribute to the software 2) Report issues or problems with the software 3) Seek support

### Software paper

- [x] **Summary:** Has a clear description of the high-level functionality and purpose of the software for a diverse, non-specialist audience been provided?
- [x] **A statement of need:** Do the authors clearly state what problems the software is designed to solve and who the target audience is?
- [x] **State of the field:** Do the authors describe how this software compares to other commonly-used packages?
- [x] **Quality of writing:** Is the paper well written (i.e., it does not require editing for structure, language, or writing quality)?
- [x] **References:** Is the list of references complete, and is everything cited appropriately that should be cited (e.g., papers, datasets, software)? Do references in the text use the proper [citation syntax]( https://rmarkdown.rstudio.com/authoring_bibliographies_and_citations.html#citation_syntax)?

### Conflict of interest

- [x] I confirm that I have read the [JOSS conflict of interest (COI) policy](https://github.com/openjournals/joss/blob/master/COI.md) and that: I have no COIs with reviewing this work or that any perceived COIs have been waived by JOSS for the purpose of this review.

### Code of Conduct

- [x] I confirm that I read and will adhere to the [JOSS code of conduct](https://joss.theoj.org/about#code_of_conduct).

### General checks

- [x] **Repository:** Is the source code for this software available at the <a target="_blank" href="https://github.com/csdms/bmi">repository url</a>?
- [x] **License:** Does the repository contain a plain-text LICENSE file with the contents of an [OSI approved](https://opensource.org/licenses/alphabetical) software license?
- [x] **Contribution and authorship:** Has the submitting author (@mdpiper) made major contributions to the software? Does the full list of paper authors seem appropriate and complete?

### Functionality

- [x] **Installation:** Does installation proceed as outlined in the documentation?
- [x] **Functionality:** Have the functional claims of the software been confirmed?
- [x] **Performance:** If there are any performance claims of the software, have they been confirmed? (If there are no claims, please check off this item.)

### Documentation

- [x] **A statement of need:** Do the authors clearly state what problems the software is designed to solve and who the target audience is?
- [x] **Installation instructions:** Is there a clearly-stated list of dependencies? Ideally these should be handled with an automated package management solution.
- [x] **Example usage:** Do the authors include examples of how to use the software (ideally to solve real-world analysis problems).
- [x] **Functionality documentation:** Is the core functionality of the software documented to a satisfactory level (e.g., API method documentation)?
- [x] **Automated tests:** Are there automated tests or manual steps described so that the functionality of the software can be verified?
- [x] **Community guidelines:** Are there clear guidelines for third parties wishing to 1) Contribute to the software 2) Report issues or problems with the software 3) Seek support

### Software paper

- [x] **Summary:** Has a clear description of the high-level functionality and purpose of the software for a diverse, non-specialist audience been provided?
- [x] **A statement of need:** Do the authors clearly state what problems the software is designed to solve and who the target audience is?
- [x] **State of the field:** Do the authors describe how this software compares to other commonly-used packages?
- [x] **Quality of writing:** Is the paper well written (i.e., it does not require editing for structure, language, or writing quality)?
- [x] **References:** Is the list of references complete, and is everything cited appropriately that should be cited (e.g., papers, datasets, software)? Do references in the text use the proper [citation syntax]( https://rmarkdown.rstudio.com/authoring_bibliographies_and_citations.html#citation_syntax)?

**Comment by whedon** [comment]
Hello human, I'm @whedon, a robot that can help you with some common editorial tasks. @yangbai90, @teuben it looks like you're currently assigned to review this paper :tada:.

**:warning: JOSS reduced service mode :warning:**

Due to the challenges of the COVID-19 pandemic, JOSS is currently operating in a "reduced service mode". You can read more about what that means in [our blog post](https://blog.joss.theoj.org/2020/05/reopening-joss).

**:star: Important :star:**

If you haven't already, you should seriously consider unsubscribing from GitHub notifications for this (https://github.com/openjournals/joss-reviews) repository. As a reviewer, you're probably currently watching this repository which means for GitHub's default behaviour you will receive notifications (emails) for all reviews 😿

To fix this do the following two things:

1. Set yourself as 'Not watching' https://github.com/openjournals/joss-reviews:

![watching](https://cloud.githubusercontent.com/assets/4483/20250593/64d7ce48-a9de-11e6-9225-d3dfb3e48e68.png)

2. You may also like to change your default settings for this watching repositories in your GitHub profile here: https://github.com/settings/notifications

![notifications](https://cloud.githubusercontent.com/assets/4483/20250604/79eb3144-a9de-11e6-9e51-0fe940966689.png)


For a list of things I can do to help you, just type:

```
@whedon commands
```

For example, to regenerate the paper pdf after making changes in the paper's md or bib files, type:

```
@whedon generate pdf
```

**Comment by whedon** [comment]
```
Reference check summary:

OK DOIs

- 10.1029/2018JF004757 is OK

MISSING DOIs

- https://doi.org/10.1177/1094342011414036 may be missing for title: High-performance language interoperability for scientific computing through Babel
- https://doi.org/10.5194/nhess-19-1723-2019 may be missing for title: Evaluating the impact of model complexity on flood wave propagation and inundation extent with a hydrologic-hydrodynamic model coupling framework
- https://doi.org/10.1016/j.cageo.2012.04.002 may be missing for title: A component-based approach to integrated modeling in the geosciences: The design of CSDMS

INVALID DOIs

- None
```

**Comment by whedon** [comment]
[ :point_right: Check article proof :page_facing_up: :point_left: ](https://github.com/openjournals/joss-papers/blob/joss.02317/joss.02317/10.21105.joss.02317.pdf)

**Comment by mdpiper** [comment]
A quick note: in a pre-submission [issue](https://github.com/openjournals/joss/issues/690), editors suggested this submission be reviewed in the context of the language specifications and examples derived from the interface defined in the main repository.

**Comment by diehlpk** [comment]
@mdpiper Are these listed in the JOSS paper as well? If, not would it be possible to add them?

**Comment by mdpiper** [comment]
@diehlpk Yup! See Table 1 in the paper.

**Comment by diehlpk** [comment]
@yangbai90 Can you please let me know if you finished your review?

**Comment by yangbai90** [comment]
Hi @diehlpk, on my side, the paper is ready for publishing.

**Comment by diehlpk** [comment]
Hi @yangbai90 thanks for your effort supporting JOSS :)

**Comment by diehlpk** [comment]
@teuben How is your review going?

**Comment by teuben** [comment]
I hope to finish this by tomorrow.

**Comment by diehlpk** [comment]
@teuben Thanks for the update.

**Comment by mdpiper** [comment]
Hi @teuben, I think we've addressed your comments in https://github.com/csdms/bmi/issues/68, https://github.com/csdms/bmi/issues/69, and https://github.com/csdms/bmi/issues/70. When you have time, would you please check the open PRs listed under each issue? If they pass muster, I'll merge the PRs and close the issues.

**Comment by teuben** [comment]
so done!  I am finished with my review (with my apologies as google decided to count all my joss emails as spam)

**Comment by diehlpk** [comment]
@teuben Thanks.

**Comment by diehlpk** [comment]
@mdpiper I will do the final review on the paper by the end of this week. After that

`When a submission is ready to be accepted, we ask that the authors issue a new tagged release of the software (if changed), and archive it (on Zenodo, figshare, or other). The authors then post the version number and archive DOI in the REVIEW issue. `

**Comment by diehlpk** [comment]
@mdpiper 

`State of the field: Do the authors describe how this software compares to other commonly-used packages?`

The reviewers agreed on this, however, I am as not as deeply in the topic would prefer if you can highlight this point in the paper a little more. So a broader audience can see this point.

Everything else is fine.

**Comment by mdpiper** [comment]
@diehlpk Thanks for your editorial review. I'll ask some text to this point. And another thank you--this a good point to make considering model coupling is decades old in, e.g., atmospheric sciences, but still relatively new in the geosciences.

**Comment by diehlpk** [comment]
@whedon generate pdf

**Comment by whedon** [comment]
[ :point_right: Check article proof :page_facing_up: :point_left: ](https://github.com/openjournals/joss-papers/blob/joss.02317/joss.02317/10.21105.joss.02317.pdf)

**Comment by mdpiper** [comment]
@diehlpk Waiting on my coauthors to review my changes...

**Comment by diehlpk** [comment]
@mdpiper Any news from your crew?

**Comment by mdpiper** [comment]
@diehlpk Waiting on one. We have a meeting this aft; I'll nudge him then.

**Comment by mdpiper** [comment]
@diehlpk Nudging successful! The updated paper text, with a broader introduction, is in the mdpiper/paper branch. If you're OK with the changes, I'll follow your instructions upthread for finalizing the submission.

**Comment by mdpiper** [comment]
@whedon generate pdf

**Comment by whedon** [comment]
[ :point_right: Check article proof :page_facing_up: :point_left: ](https://github.com/openjournals/joss-papers/blob/joss.02317/joss.02317/10.21105.joss.02317.pdf)

**Comment by mdpiper** [comment]
@whedon generate pdf from branch mdpiper/paper

**Comment by whedon** [comment]
```
Attempting PDF compilation from custom branch mdpiper/paper. Reticulating splines etc...
```

**Comment by whedon** [comment]
[ :point_right: Check article proof :page_facing_up: :point_left: ](https://github.com/openjournals/joss-papers/blob/joss.02317/joss.02317/10.21105.joss.02317.pdf)

**Comment by diehlpk** [comment]
@mdpiper The addition to the paper is sufficient and I can recommend the paper for acceptance. 

Next steps are: 

`
When a submission is ready to be accepted, we ask that the authors issue a new tagged release of the software (if changed), and archive it (on Zenodo, figshare, or other). The authors then post the version number and archive DOI in the REVIEW issue.`

**Comment by mdpiper** [comment]
@diehlpk version: [v2.0](https://github.com/csdms/bmi/releases/tag/v2.0), doi: 10.5281/zenodo.3955010

On Zenodo: [Basic Model Interface 2.0](https://zenodo.org/record/3955010#.XxcnjR1MEbI)

**Comment by diehlpk** [comment]
@whedon generate pdf

**Comment by diehlpk** [comment]
@whedon check references

**Comment by whedon** [comment]
PDF failed to compile for issue #2317 with the following error: 

 sh: 0: getcwd() failed: No such file or directory
pandoc: 10.21105.joss.02317.pdf: openBinaryFile: does not exist (No such file or directory)
Looks like we failed to compile the PDF

**Comment by whedon** [comment]
```
Reference check summary:

OK DOIs

- 10.1029/2018JF004757 is OK

MISSING DOIs

- https://doi.org/10.1177/1094342005056120 may be missing for title: Design and implementation of components in the Earth System Modeling Framework
- https://doi.org/10.1177/1094342011414036 may be missing for title: High-performance language interoperability for scientific computing through Babel
- https://doi.org/10.5194/nhess-19-1723-2019 may be missing for title: Evaluating the impact of model complexity on flood wave propagation and inundation extent with a hydrologic-hydrodynamic model coupling framework
- https://doi.org/10.1016/j.cageo.2012.04.002 may be missing for title: A component-based approach to integrated modeling in the geosciences: The design of CSDMS

INVALID DOIs

- None
```

**Comment by diehlpk** [comment]
@whedon set 10.5281/zenodo.3955010 as archive

**Comment by whedon** [comment]
OK. <a href="https://doi.org/10.5281/zenodo.3955010" target="_blank">10.5281/zenodo.3955010</a> is the archive.

**Comment by diehlpk** [comment]
@whedon set v2.0 as version

**Comment by whedon** [comment]
OK. v2.0 is the version.

**Comment by diehlpk** [comment]
@whedon generate pdf from branch mdpiper/paper

**Comment by whedon** [comment]
```
Attempting PDF compilation from custom branch mdpiper/paper. Reticulating splines etc...
```

**Comment by whedon** [comment]
[ :point_right: Check article proof :page_facing_up: :point_left: ](https://github.com/openjournals/joss-papers/blob/joss.02317/joss.02317/10.21105.joss.02317.pdf)

**Comment by diehlpk** [comment]
@mdpiper Can you please check this version of the paper as the final version? After you confirmed I will pass the paper to the EICs. We like to avoid to do some after publication changes.

**Comment by mdpiper** [comment]
@diehlpk The rendered PDF is correct; however, I merged the mdpiper/paper branch into master earlier this morning. (I'm not sure if this is important, but I want to check.)

**Comment by diehlpk** [comment]
@whedon accept

**Comment by whedon** [comment]
```
Attempting dry run of processing paper acceptance...
```

**Comment by whedon** [comment]
```
Reference check summary:

OK DOIs

- 10.1029/2018JF004757 is OK

MISSING DOIs

- https://doi.org/10.1177/1094342005056120 may be missing for title: Design and implementation of components in the Earth System Modeling Framework
- https://doi.org/10.1177/1094342011414036 may be missing for title: High-performance language interoperability for scientific computing through Babel
- https://doi.org/10.5194/nhess-19-1723-2019 may be missing for title: Evaluating the impact of model complexity on flood wave propagation and inundation extent with a hydrologic-hydrodynamic model coupling framework
- https://doi.org/10.1016/j.cageo.2012.04.002 may be missing for title: A component-based approach to integrated modeling in the geosciences: The design of CSDMS

INVALID DOIs

- None
```

**Comment by whedon** [comment]
:wave: @openjournals/joss-eics, this paper is ready to be accepted and published.

 Check final proof :point_right: https://github.com/openjournals/joss-papers/pull/1586

If the paper PDF and Crossref deposit XML look good in https://github.com/openjournals/joss-papers/pull/1586, then you can now move forward with accepting the submission by compiling again with the flag `deposit=true` e.g.
 ```
@whedon accept deposit=true
```

**Comment by mdpiper** [comment]
Thanks, @diehlpk, @yangbai90, and @teuben!

**Comment by Kevin-Mattheus-Moerman** [comment]
@mdpiper I've reviewed your paper. It is ready for acceptance once you make the below change. You can run `@whedon generate pdf` here to update the paper once you are finished. 

- [ ] Can you add countries to your affiliations?

**Comment by mdpiper** [comment]
@Kevin-Mattheus-Moerman Thank you! Done!

**Comment by mdpiper** [comment]
@whedon generate pdf

**Comment by whedon** [comment]
[ :point_right: Check article proof :page_facing_up: :point_left: ](https://github.com/openjournals/joss-papers/blob/joss.02317/joss.02317/10.21105.joss.02317.pdf)

**Comment by Kevin-Mattheus-Moerman** [comment]
@whedon accept

**Comment by whedon** [comment]
```
Attempting dry run of processing paper acceptance...
```

**Comment by whedon** [comment]
```
Reference check summary:

OK DOIs

- 10.1029/2018JF004757 is OK

MISSING DOIs

- https://doi.org/10.1177/1094342005056120 may be missing for title: Design and implementation of components in the Earth System Modeling Framework
- https://doi.org/10.1177/1094342011414036 may be missing for title: High-performance language interoperability for scientific computing through Babel
- https://doi.org/10.5194/nhess-19-1723-2019 may be missing for title: Evaluating the impact of model complexity on flood wave propagation and inundation extent with a hydrologic-hydrodynamic model coupling framework
- https://doi.org/10.1016/j.cageo.2012.04.002 may be missing for title: A component-based approach to integrated modeling in the geosciences: The design of CSDMS

INVALID DOIs

- None
```

**Comment by whedon** [comment]
:wave: @openjournals/joss-eics, this paper is ready to be accepted and published.

 Check final proof :point_right: https://github.com/openjournals/joss-papers/pull/1587

If the paper PDF and Crossref deposit XML look good in https://github.com/openjournals/joss-papers/pull/1587, then you can now move forward with accepting the submission by compiling again with the flag `deposit=true` e.g.
 ```
@whedon accept deposit=true
```

**Comment by Kevin-Mattheus-Moerman** [comment]
@whedon accept deposit=true

**Comment by whedon** [comment]
```
Doing it live! Attempting automated processing of paper acceptance...
```

**Comment by whedon** [comment]
🐦🐦🐦 👉 [Tweet for this paper](https://twitter.com/JOSS_TheOJ/status/1286349291682967560) 👈 🐦🐦🐦

**Comment by whedon** [comment]
🚨🚨🚨 **THIS IS NOT A DRILL, YOU HAVE JUST ACCEPTED A PAPER INTO JOSS!** 🚨🚨🚨

 Here's what you must now do:

0. Check final PDF and Crossref metadata that was deposited :point_right: https://github.com/openjournals/joss-papers/pull/1588
1. Wait a couple of minutes to verify that the paper DOI resolves [https://doi.org/10.21105/joss.02317](https://doi.org/10.21105/joss.02317)
2. If everything looks good, then close this review issue.
3. Party like you just published a paper! 🎉🌈🦄💃👻🤘

 Any issues? Notify your editorial technical team...

**Comment by whedon** [comment]
:tada::tada::tada: Congratulations on your paper acceptance! :tada::tada::tada:

If you would like to include a link to your paper from your README use the following code snippets:

```
Markdown:
[![DOI](https://joss.theoj.org/papers/10.21105/joss.02317/status.svg)](https://doi.org/10.21105/joss.02317)

HTML:
<a style="border-width:0" href="https://doi.org/10.21105/joss.02317">
  <img src="https://joss.theoj.org/papers/10.21105/joss.02317/status.svg" alt="DOI badge" >
</a>

reStructuredText:
.. image:: https://joss.theoj.org/papers/10.21105/joss.02317/status.svg
   :target: https://doi.org/10.21105/joss.02317
```
This is how it will look in your documentation:

[![DOI](https://joss.theoj.org/papers/10.21105/joss.02317/status.svg)](https://doi.org/10.21105/joss.02317)

**We need your help!**

Journal of Open Source Software is a community-run journal and relies upon volunteer effort. If you'd like to support us please consider doing either one (or both) of the the following:

- Volunteering to review for us sometime in the future. You can add your name to the reviewer list here: https://joss.theoj.org/reviewer-signup.html
- Making a small donation to support our running costs here: https://numfocus.org/donate-to-joss

**Comment by mdpiper** [comment]
Yay! cc: @mcflugen @gregtucker

**Comment by mdpiper** [comment]
@diehlpk I noticed in the citation string on the sidebar that the volume/issue number on this paper is 1(1), but the papers before/after it are 5(51). (And in the PDF of this paper, the volume/issue are listed as 5(51)) Can you recommend where to check on or report this?

**Comment by arfon** [comment]
> @diehlpk I noticed in the citation string on the sidebar that the volume/issue number on this paper is 1(1), but the papers before/after it are 5(51). (And in the PDF of this paper, the volume/issue are listed as 5(51)) Can you recommend where to check on or report this?

I've just fixed this. Not quite sure what happened here...

**Comment by mdpiper** [comment]
@arfon Thank you!
